okhttp/src/main/java/okhttp3/OkHttpClient.kt

```diff
@@ -0,0 +1,829 @@
+ * do so.
+ *
+ * Shutdown the dispatcher's executor service with [shutdown()][ExecutorService.shutdown].
+ * This will also cause future calls to the client to be rejected.
+ * ```
+ * client.dispatcher().executorService().shutdown();
+ * ```
+ *
+ * Clear the connection pool with [evictAll()][ConnectionPool.evictAll]. Note that the
+ * connection pool's daemon thread may not exit immediately.
+ * ```
+ * client.connectionPool().evictAll();
+ * ```
+ *
+ *
+ * If your client has a cache, call [close()][Cache.close]. Note that it is an error to
+ * create calls against a cache that is closed, and doing so will cause the call to crash.
+ * ```
+ * client.cache().close();
+ * ```
+ *
+ * OkHttp also uses daemon threads for HTTP/2 connections. These will exit automatically if they
+ * remain idle.
+ */
+class OkHttpClient internal constructor(
+  builder: Builder
+) : Cloneable, Call.Factory, WebSocket.Factory {
+
+  private val dispatcher: Dispatcher = builder.dispatcher
+  private val proxy: Proxy? = builder.proxy
+  private val protocols: List<Protocol> = builder.protocols
+  private val connectionSpecs: List<ConnectionSpec> = builder.connectionSpecs
+  private val interceptors: List<Interceptor> = Util.immutableList(builder.interceptors)
+  private val networkInterceptors: List<Interceptor> = Util.immutableList(builder.networkInterceptors)
+  private val eventListenerFactory: EventListener.Factory = builder.eventListenerFactory
+  private val proxySelector: ProxySelector = builder.proxySelector
+  private val cookieJar: CookieJar = builder.cookieJar
+  private val cache: Cache? = builder.cache
+  private val socketFactory: SocketFactory = builder.socketFactory
+  private val sslSocketFactory: SSLSocketFactory
+  private val hostnameVerifier: HostnameVerifier = builder.hostnameVerifier
+  private val certificatePinner: CertificatePinner
+  private val proxyAuthenticator: Authenticator = builder.proxyAuthenticator
+  private val authenticator: Authenticator = builder.authenticator
+  private val connectionPool: ConnectionPool = builder.connectionPool
+  private val dns: Dns = builder.dns
+  private val followSslRedirects: Boolean = builder.followSslRedirects
+  private val followRedirects: Boolean = builder.followRedirects
+  private val retryOnConnectionFailure: Boolean = builder.retryOnConnectionFailure
+  private val callTimeout: Int = builder.callTimeout
+  private val connectTimeout: Int = builder.connectTimeout
+  private val readTimeout: Int = builder.readTimeout
+  private val writeTimeout: Int = builder.writeTimeout
+  private val pingInterval: Int = builder.pingInterval
+  private val internalCache: InternalCache? = builder.internalCache
+  private val certificateChainCleaner: CertificateChainCleaner
+
+  constructor() : this(Builder())
+
+  init {
+    var isTLS = false
+    for (spec in connectionSpecs) {
+      isTLS = isTLS || spec.isTls
+    }
+
+    // TODO - these nullability warnings existed in Java
```

Comment:
what’s actionable here? My preference is to just omit the TODO

Reply:
I think it's a legit issue.

Either sslSocketFactory should be nullable here https://github.com/square/okhttp/blob/master/okhttp/src/main/java/okhttp3/OkHttpClient.java#L327

Because this line could possibly assign a nullable value to it even if the builder value is `null` if `isTLS` is `false`.

```kotlin
if (builder.sslSocketFactory != null || !isTLS) {
  this.sslSocketFactory = builder.sslSocketFactory
```

We can leave it with just the force-unwraps here, but the null-check below it suggested to me what it was something intended to be nullable. What do you think?

```kotlin
if (sslSocketFactory != null) {
  Platform.get().configureSslSocketFactory(sslSocketFactory)
}
```

And in either case, `builder.certificateChainCleaner` is implicitly assumed to be present and subsequently force-unwrapped.

Reply:
I think it's too strict as-is. Previously a null value would cause a theoretical NPE after calling the accessor; now you'll always get an NPE immediately when you construct.

Fix? Probably to make accessing this value throw if there's no TLS.

Reply:
> Probably to make accessing this value throw if there's no TLS.

I think I didn't quite follow what `this` means in this context?

Reply:
I'd like to make the accessors return non-null values. If you attempt to read either on a client that doesn't have SSL, it should throw a UnsupportedOperationException.